### DISCUSSION TITLE
How to use `NOT EXISTS` in filter?

### DISCUSSION THREAD

**Discussion by lz1998** [SeaQL/sea-orm#982] [Q&A]
Database：MySQL
| Name | Type |
|---|---|
| id | integer |
| vid | string |
| status | integer  |

```sql
SELECT 
DISTINCT(vid) 
FROM project a 
WHERE vid IN ('aaa','bbb') 
AND NOT EXISTS (
    SELECT*FROM project WHERE status!=1 AND vid=a.vid
)
```

'aaa', 'bbb' is a `Vec<String>`. I have tried `from_sql_and_values` and put `vec![vids.into()]` in the second param, and two errors occur:
```rust
    let vids=vec!["aaa".to_string(),"bbb".to_string()];
    let result: Vec<String> = Entity::find()
    .from_raw_sql(Statement::from_sql_and_values(
        DbBackend::MySql,
        r#"SELECT 
        DISTINCT(vid) 
        FROM project a 
        WHERE vid IN ('aaa','bbb') 
        AND NOT EXISTS (
        SELECT*FROM project WHERE status!=1 AND vid=a.vid
        )"#,
        vec![vids.into()],
    ))
    .all(conn)
    .await.unwrap();
```

```bash
the trait `From<Vec<std::string::String>>` is not implemented for `sea_orm::Value`

expected struct `std::string::String`, found struct `database::project::Model`
```

**Comment by ikrivosheev** [comment]
@lz1998, hello! Maybe the answer is not for your question but I prepare PR for support `EXISTS` in SeaQuery: https://github.com/SeaQL/sea-query/pull/379. After there will be native support in SeaORM.

**Reply by lz1998** [reply]
If I use `from_sql_and_values` instead of `exists`, how to solve the two problems?

**Reply by billy1624** [reply]
Hey @lz1998, then you'd have to manually construct the SQL with placeholder for value binding.

```rs
let result: Vec<String> = Entity::find()
    .from_raw_sql(Statement::from_sql_and_values(
        DbBackend::MySql,
        r#"SELECT 
        DISTINCT(vid) 
        FROM project a 
        WHERE vid IN (?, ?) 
        AND NOT EXISTS (
        SELECT*FROM project WHERE status!=1 AND vid=a.vid
        )"#,
        vec![
            "aaa".into(),
            "bbb".into(),
        ],
    ))
    .all(conn)
    .await
    .unwrap();
```

**Reply by lz1998** [reply]
> Hey @lz1998, then you'd have to manually construct the SQL with placeholder for value binding.
> 
> ```rust
> let result: Vec<String> = Entity::find()
>     .from_raw_sql(Statement::from_sql_and_values(
>         DbBackend::MySql,
>         r#"SELECT 
>         DISTINCT(vid) 
>         FROM project a 
>         WHERE vid IN (?, ?) 
>         AND NOT EXISTS (
>         SELECT*FROM project WHERE status!=1 AND vid=a.vid
>         )"#,
>         vec![
>             "aaa".into(),
>             "bbb".into(),
>         ],
>     ))
>     .all(conn)
>     .await
>     .unwrap();
> ```

The length of vids is not always 2, do I need to generate the SQL with placeholder according to the length of vids?

**Reply by billy1624** [reply]
Yeah, that's true.

Why don't we rewrite the SQL into...

```sql
SELECT DISTINCT(vid) 
FROM project
WHERE
    vid IN ('aaa','bbb')
    AND vid NOT IN (SELECT vid FROM project WHERE status != 1)
```

**Reply by lz1998** [reply]
> Yeah, that's true.
> 
> Why don't we rewrite the SQL into...
> 
> ```sql
> SELECT DISTINCT(vid) 
> FROM project
> WHERE
>     vid IN ('aaa','bbb')
>     AND vid NOT IN (SELECT vid FROM project WHERE status != 1)
> ```

Then how to get a `Vec<String>` with raw_sql?
```bash
expected struct `std::string::String`, found struct `database::project::Model`
```

**Reply by lz1998** [reply]
With `Result` struct, it works but seems strange.
```rust
    #[derive(FromQueryResult)]
    struct Result {
        pub vid: String,
    }
    Entity::find()
        .from_raw_sql(sql)
        .into_model::<Result>()
        .all(conn)
        .await
        .map(|m| m.into_iter().map(|p| p.vid).collect())
        .map_err(Into::into)
```

**Reply by billy1624** [reply]
Hey @lz1998, this SQL can be constructed with SeaQuery.

```rs
SELECT DISTINCT(vid) 
FROM project
WHERE
    vid IN ('aaa','bbb')
    AND vid NOT IN (SELECT vid FROM project WHERE status != 1)
```
